File: tools/sprites/backink.mjs
/**
 * 帰り道（`.chain-foot a` / `.chap-note a`）の**字の濃さ**を、描かれた画素から測る。
 *
 *   SPORT=4210 W=390 TAG=before node tools/sprites/backink.mjs
 *
 * やり方は `pcink.mjs` と同じ。1枚そのまま撮り、2枚目は**その字だけ透明**にして
 * 撮り、差の出た画素を「字」として、その下の地と比べる。縁取りも影も字と
 * いっしょに消すので、比を水増ししない。読むのは `backink.py`。
 *
 * `pcink.mjs` と違うのは2つ。
 *
 *  1. **丸ごと撮らない。見えている窓だけ撮る。** `/island/caucasus/streams` は
 *     390px で 48,037px あり、dpr2 の丸ごとは 96,074px になって**絵の端が
 *     白く抜ける**（`docs/island-misses.md` #80 の追記）。
 *  2. **落ちたものだけでなく、測った全部の値を出す。** ここは「前と後で
 *     変わっていないこと」を見る道具なので、通ったものの値も要る。
 *
 * **dpr は 2 以上で撮る。** dpr1 だと同じ字が 2〜3割低く出る。
 * **合否は中央値で決める。**
 */
import { chromium } from "playwright-core";
import { offline } from "./route.mjs";
import { mkdirSync, writeFileSync, unlinkSync } from "fs";
import { execFileSync } from "child_process";

const SPORT = process.env.SPORT || "4210";
const W = Number(process.env.W || 390);
const H = Number(process.env.H || 900);
const DPR = Number(process.env.DPR || 2);
const TAG = process.env.TAG || "ink";
const OUT = process.env.OUT || `/tmp/backreport/ink-${TAG}-${W}`;
const SEL = process.env.SEL || ".chain-foot a,.chap-note a";
const PAGES = (
  process.env.PAGES ||
  "/atlas," +
    "/island/albania,/island/caucasus,/island/europe,/island/iran-walk,/island/middle-east,/island/nordic," +
    "/island/caucasus/streams,/island/europe/streams,/island/iran-walk/streams,/island/middle-east/streams"
).split(",");

mkdirSync(OUT, { recursive: true });
const b = await chromium.launch({
  executablePath: "/opt/pw-browsers/chromium-1194/chrome-linux/chrome",
  args: ["--no-sandbox"],
});
const ctx = await b.newContext({
  viewport: { width: W, height: H },
  deviceScaleFactor: DPR,
  isMobile: W < 900,
  hasTouch: W < 900,
  reducedMotion: "reduce",
});
await offline(ctx);
await ctx.addInitScript(() => {
  try {
    localStorage.setItem("ayato-island-arrived", "2026-09-04");
    localStorage.setItem("ayato-island-walked", "1");
  } catch {}
});
/* **島の地の色は時刻で変わる**（`[data-time]`）。前と後を別の時間に測ると、
   直していないものが動く（`docs/island-misses.md` #81）。帯を固定する。 */
const TIME = process.env.TIME || "day";
const p = await ctx.newPage();

const all = [];
for (const path of PAGES) {
  await p.goto(`http://localhost:${SPORT}${path === "/" ? "/index" : path}.html`, {
    waitUntil: "domcontentloaded",
    timeout: 60000,
  });
  await p.waitForTimeout(800);
  await p.evaluate((t) => { document.documentElement.dataset.time = t; }, TIME);
  /* **止めてから撮る。** 2枚のあいだに何かが動くと、差分に字と関係ない画素が
     混ざる。CSS の animation を止めるだけでは足りない（島は rAF で動く）。 */
  await p.evaluate(() => {
    window.requestAnimationFrame = () => 0;
    const st = document.createElement("style");
    st.textContent = "*,*::before,*::after{animation-play-state:paused!important;transition:none!important}";
    document.head.appendChild(st);
  });

  const links = await p.$$(SEL);
  for (let i = 0; i < links.length; i++) {
    const el = links[i];
    await el.evaluate((e) => e.scrollIntoView({ block: "center" }));
    await p.waitForTimeout(200);
    const box = await el.evaluate((e) => {
      const r = e.getBoundingClientRect();
      return { x: r.x, y: r.y, w: r.width, h: r.height, t: (e.textContent || "").trim().slice(0, 24),
        c: typeof e.className === "string" ? e.className : "", size: getComputedStyle(e).fontSize };
    });
    const name = `${path.replace(/\//g, "_").replace(/^_/, "") || "index"}-${i}`;
    const base = `${OUT}/${name}`;
    writeFileSync(`${base}.json`, JSON.stringify({ dpr: DPR, boxes: [box] }));
    await p.screenshot({ path: `${base}.shot.png` });
    await el.evaluate((e) => {
      e.style.setProperty("color", "transparent", "important");
      e.style.setProperty("text-shadow", "none", "important");
      e.style.setProperty("-webkit-text-stroke-color", "transparent", "important");
      // 下線も字の一部。残すと「地」に混ざる
      e.style.setProperty("text-decoration-color", "transparent", "important");
    });
    await p.waitForTimeout(120);
    await p.screenshot({ path: `${base}.bg.png` });
    await el.evaluate((e) => {
      e.style.removeProperty("color");
      e.style.removeProperty("text-shadow");
      e.style.removeProperty("-webkit-text-stroke-color");
      e.style.removeProperty("text-decoration-color");
    });

    const res = JSON.parse(
      execFileSync("python3", ["/home/user/live-streaming/tools/sprites/backink.py", base], {
        encoding: "utf8",
        maxBuffer: 1 << 26,
      }),
    )[0];
    unlinkSync(`${base}.shot.png`);
    unlinkSync(`${base}.bg.png`);
    all.push({ path, w: W, ...res });
    console.log(
      `${path}  ${W}px  「${res.t}」  ${res.why ?? `中央値 ${res.mid}  暗い地 ${res.lo}  明るい地 ${res.hi}  字 ${res.ink} / 地 ${res.bg}  ${res.mid >= 4.5 && res.lo >= 4.5 ? "OK" : "** 4.5割れ **"}`}`,
    );
  }
}
writeFileSync(`${OUT}/ink.json`, JSON.stringify(all, null, 1));
const bad = all.filter((r) => !r.why && !(r.mid >= 4.5 && r.lo >= 4.5));
console.log(`\n測った ${all.length}か所 / 4.5割れ ${bad.length}か所  →  ${OUT}/ink.json`);
await b.close();

File: tools/sprites/backink.py
"""帰り道の字の濃さを、描かれた画素から測る（`backink.mjs` が撮った2枚を読む）。

**計算は `pcink.py`（＝`inkpx.py`）と同じもの。1行も変えていない。**
違うのは2つだけ。

  1. **落ちたものだけでなく、測った全部を返す。** ここは「前と後で変わって
     いないこと」を見る道具なので、4.5 を越えているものの値も要る
  2. **丸ごと撮った絵を読まない。** `/island/caucasus/streams` は 390px で
     48,037px あり、dpr2 の丸ごとは 96,074px になって**絵の端が白く抜ける**
     （`docs/island-misses.md` #80 の追記）。抜けた白を地として数えると、
     読める字が「割れている」と出る。だから見えている窓だけを撮って、
     窓の中の座標で読む

    使い方: python3 backink.py <base(.json/.shot.png/.bg.png の手前まで)>
"""
import json
import sys

import numpy as np
from PIL import Image

base = sys.argv[1]

d = json.load(open(base + ".json"))
dpr = d["dpr"]
shot = np.asarray(Image.open(base + ".shot.png").convert("RGB"), dtype=np.int16)
bg = np.asarray(Image.open(base + ".bg.png").convert("RGB"), dtype=np.int16)
H, W = shot.shape[:2]


def lum_a(px):
    c = px.astype(np.float64) / 255.0
    c = np.where(c <= 0.04045, c / 12.92, ((c + 0.055) / 1.055) ** 2.4)
    return c[..., 0] * 0.2126 + c[..., 1] * 0.7152 + c[..., 2] * 0.0722


def lin(c):
    c /= 255
    return c / 12.92 if c <= 0.04045 else ((c + 0.055) / 1.055) ** 2.4


def lum(rgb):
    r, g, b = rgb
    return 0.2126 * lin(r) + 0.7152 * lin(g) + 0.0722 * lin(b)


def ratio(a, b):
    la, lb = lum(a), lum(b)
    if la < lb:
        la, lb = lb, la
    return (la + 0.05) / (lb + 0.05)


out = []
for b0 in d["boxes"]:
    x0, y0 = max(0, int(b0["x"] * dpr)), max(0, int(b0["y"] * dpr))
    x1 = min(W, int((b0["x"] + b0["w"]) * dpr))
    y1 = min(H, int((b0["y"] + b0["h"]) * dpr))
    if x1 - x0 < 2 or y1 - y0 < 2:
        out.append({"t": b0["t"], "why": "窓の外"})
        continue
    a = shot[y0:y1, x0:x1]
    c = bg[y0:y1, x0:x1]
    mask = np.abs(a - c).sum(axis=2) > 40
    if int(mask.sum()) < 6:
        out.append({"t": b0["t"], "why": "字の画素が見つからない"})
        continue
    unders = c[mask]
    diff = np.abs(a - c).sum(axis=2)[mask]
    core = diff >= np.quantile(diff, 0.6)
    use_core = int(core.sum()) >= 4
    painted = a[mask][core] if use_core else a[mask]
    lt = lum_a(painted)
    lb = lum_a(unders[core] if use_core else unders)
    rs = (np.maximum(lt, lb) + 0.05) / (np.minimum(lt, lb) + 0.05)
    mid = float(np.median(rs))
    ink = tuple(int(v) for v in np.median(painted, axis=0))
    ub = unders[core] if use_core else unders
    order = np.argsort(lum_a(ub))
    lo = tuple(int(v) for v in ub[order[len(order) // 20]])
    hi = tuple(int(v) for v in ub[order[-1 - len(order) // 20]])
    out.append({
        "t": b0["t"], "c": b0["c"][:40], "size": b0["size"],
        "mid": round(mid, 2), "lo": round(ratio(ink, lo), 2), "hi": round(ratio(ink, hi), 2),
        "ink": "#%02x%02x%02x" % ink, "bg": "#%02x%02x%02x" % lo,
        "px": int(mask.sum()),
    })

print(json.dumps(out, ensure_ascii=False))
